src/wandb_osh/syncer.py

```diff
@@ -44,44 +66,42 @@ def sync(self, dir: PathLike) -> None:
         """
         sync_dir(dir, options=self.wandb_options, timeout=self._timeout)
 
-    def loop(self) -> None:
+    def dir_watcher(self) -> None:
         """Read command files and trigger syncing"""
         logger.info(
             "wandb-osh v%s, starting to watch %s", __version__, self.command_dir
         )
         while True:
             start_time = time.time()
             self.command_dir.mkdir(parents=True, exist_ok=True)
-            command_files = []
-            targets = []
             for command_file in self.command_dir.glob("*.command"):
                 target = Path(command_file.read_text())
-                command_files.append(command_file)
                 if not target.is_dir():
                     logger.error(
                         "Command file %s points to non-existing directory %s",
                         command_file,
                         target,
                     )
                     continue
-                targets.append(target)
-            for target in set(targets):
-                logger.info("Syncing %s...", target)
-                try:
-                    self.sync(target)
-                except subprocess.TimeoutExpired:
-                    # try again later
-                    logger.warning("Syncing %s timed out. Trying later.", target)
-                    from wandb_osh.hooks import TriggerWandbSyncHook
-
-                    TriggerWandbSyncHook(self.command_dir)(target)
-            time.sleep(0.25)
-            for cf in command_files:
+                self.target_queue.put((command_file, target))
+            time.sleep(max(0.0, (time.time() - start_time) - self.wait))
+
+    def worker(self) -> None:
+        while True:
+            try:
+                cf, target = self.target_queue.get(timeout=self._timeout)
+                self.sync(target)
+                time.sleep(0.25)
                 if cf.is_file():
                     cf.unlink()
-            if "PYTEST_CURRENT_TEST" in os.environ:
-                break
-            time.sleep(max(0.0, (time.time() - start_time) - self.wait))
+                if "PYTEST_CURRENT_TEST" in os.environ:
+                    break
+            except Empty:
```

Comment:
This seems to be for the case that we haven't encountered any new synchronization requests. So this should be after `target_queue.get` and simply be ignored (with a small timeout), right? Because it's nothing bad.

The "try again later" code below is if `self.sync(target)` times out